src/StereoRecorder.cpp

```diff
@@ -45,61 +45,49 @@ void StereoRecorder::updateSamples()
 	// But avoid allocating an array each frame
 	static TimeBuffer timeSpace;
 
-	for (size_t k = 0; k < FFT_SIZE; ++k)
+	if (m_bufferOverflow)
 	{
-		timeSpace[k] = s_hanning[k] * m_leftData.at(k) / static_cast<FFTType>(INT16_MAX);
+		// Swap buffers
+		m_bufferMutex.lock();
+		std::swap(m_leftDataBuffer, m_leftData);
+		std::swap(m_rightDataBuffer, m_rightData);
+		m_bufferMutex.unlock();
+		m_bufferOverflow = false;
 	}
 
+	for (size_t k = 0; k < FFT_SIZE; ++k)
+		timeSpace[k] = s_hanning[k] * m_leftData.at(k) / static_cast<FFTType>(INT16_MAX);
+
 	m_FFT.computeFFT(timeSpace, m_leftFreq);
 
 	for (size_t k = 0; k < FFT_SIZE; ++k)
-	{
 		timeSpace[k] = s_hanning[k] * m_rightData.at(k) / static_cast<FFTType>(INT16_MAX);
-	}
 
 	m_FFT.computeFFT(timeSpace, m_rightFreq);
 }
 
 bool StereoRecorder::onProcessSamples(const sf::Int16* _samples, std::size_t _sampleCount)
 {
-	size_t maxSample = _sampleCount / 2;
-	if (maxSample + m_pos > 2 * FFT_SIZE)
-		maxSample = 2 * FFT_SIZE - m_pos;
-
-	bool overflow = false;
-	size_t nbSample = maxSample;
+	// Discard samples until the main thread (updateSamples) swaps buffers
+	if(m_bufferOverflow)
+		return true;
 
+	size_t nbSample = _sampleCount / 2;
 	if (m_pos + nbSample >= FFT_SIZE)
 	{
 		nbSample = FFT_SIZE - m_pos;
-		overflow = true;
+		m_bufferOverflow = true;
 	}
 
+	m_bufferMutex.lock();
 	for (size_t i = 0; i < nbSample; ++i)
 	{
 		m_leftDataBuffer[m_pos + i] = _samples[2 * i];
 		m_rightDataBuffer[m_pos + i] = _samples[2 * i + 1];
 	}
+	m_bufferMutex.unlock();
 
-	if (overflow)
-	{
-		// Swap buffers
-		m_bufferMutex.lock();
-		std::swap(m_leftDataBuffer, m_leftData);
-		std::swap(m_rightDataBuffer, m_rightData);
-		m_bufferMutex.unlock();
-
-		// Fill the remaining samples
-		for (size_t i = 0; i < maxSample - nbSample; ++i)
-		{
-			m_leftDataBuffer[i] = _samples[2 * (nbSample + i)];
-			m_rightDataBuffer[i] = _samples[2 * (nbSample + i) + 1];
-		}
-
-		m_pos = maxSample - nbSample;
-	}
-	else
-		m_pos += nbSample;
+	m_pos = m_bufferOverflow ? 0 : m_pos + nbSample;
```

Comment:
My vote for adding braces here to save readers wondering about operator precedence.

Reply:
@hartwork Done!
I've added braces around the condition too, some people prefers that also for readability 🙂

Reply:
Thanks! :+1: